### ISSUE TITLE
more malformed LAS files from USGS to handle, mostly florida

### ISSUE THREAD

**Issue by JustinGOSSES** [JustinGOSSES/wellio.js#61]
https://github.com/JustinGOSSES/wellio.js/tree/master/assets/usgs_bad

Need to get wellio to handle these. Suspect they are mostly excessive lines with hashes that are causing problems but not sure yet.

**Comment by dcslagel** [comment]
In starting to look at this, it looks like 1609901672340.las is successfully handled by wellio.js.  Although it doesn't have a UWI entry and has an empty API value it still seems to parse into JSON.   The curve graphs seem to display properly, but could you verify?

Thanks

**Comment by JustinGOSSES** [comment]
I made some changes to wellio that fix the lack of UWI in some cases since adding this issue, so it makes sense that 1609901672340.las and 1610001023940.las work now.

**Comment by dcslagel** [comment]
@JustinGOSSES ,

How or should 260341080252801.19980915.NN.las be handled? 

 Its main problem is that it has **two** '~WELL INFORMATION` headings as follows:

~Version Information 
... # version metadata rows
~WELL INFORMATION BLOCK                                                                                                                                                                                                                           
... # well metadata rows
~Curve Information Block
~WELL INFORMATION BLOCK
...  # curve metadata rows

The current behavior is that the final Well section is populated with the Curve metadata and the Curve section is empty.

We could change the code to ignore the 2nd Well header and populate the Curve section with the Curve metadata. However if  we did that and there is ever a las file with a 2nd Well block with metadata following an empty Curve block, it would parse incorrectly.

The LAS-2.0 spec say there should be only one of each of the standard sections: Version, Well, Curve, Parameter, Other and Ascii.

Would it be better to exit with a message that states the file contains 2 of a standard section header and let the person manually correct the las file?

Thanks,
DC

**Comment by JustinGOSSES** [comment]
That's a tricky one, but I'm favoring the error message.

We might reach out to lasio maintainers in the SWUNG slack as they may have already taken an approach on this problem.

I'm wondering if we might also add some optional arguments to las2json so users can specify the behavior? That might add complexity to the codebase though, which might not be worth it if problem is rare.

**Comment by dcslagel** [comment]
In Lasio,  260341080252801.19980915.NN.las has the same behavior as wellio.js: 
The final Well section is populated with the Curve metadata and the Curve section is empty.  
In addition Lasio then sets the Curve names 'UNKNOWN#':     
`['UNKNOWN:1', 'UNKNOWN:2', 'UNKNOWN:3', 'UNKNOWN:4'] `

**Comment by kinverarity1** [comment]
@JustinGOSSES I haven't seen these files before. They're awful! @dcslagel I've put a notebook of the lasio behaviour you mention here - https://gist.github.com/kinverarity1/0c3d9e34da916f2b60dafb04f58557fd

I consider this a unnoticed bug in lasio - we are dropping all of the valuable info in the first Well section for no good reason. Not sure how to handle it but I'd look for some way of at least retaining both sections... will think about it I don't like exiting with an error message if we can avoid it. Certainly a warning message and perhaps a way to combine the two Well sections.

**Comment by dcslagel** [comment]
@kinverarity1 , 

Thanks for the gist.  Yeah, it is the same as I saw, even better because it makes the missing colon in the version line visible to everyone.  We can change the code to handle the missing colon in the weillio.js/lasio codebases.

@JustinGOSSES , @kinverarity1 ,

What do you think about this approach or some variation on it?  These probably won't catch everything future case but should pass existing test cases and handle this example along with a few others.

1. Register when a section has been read.

2. If we are in the ~Curve section and see another section header before any data, check if that section header has been registered.  Then take one of the following actions...

3. If the new section has NOT been registered, then we treat the ~Curve section as empty and read following data as belonging to the new section.

4. If the new section had been previously registered and if the next bit of data is meets the standards requirements for the Curve index ("The first channel described is the index of all other channels, and must be depth, time, or index. The only valid mnemonics for the index channel are DEPT, DEPTH, TIME or INDEX."),  Then we ignore the 2nd section header and parse the data as ~Curve section data

5. If the new section had been previously registered and the next bit of data does NOT meet the standards requirements for the Curve index, then we treat the ~Curve section as empty and read following data as belonging to the new section.

**Comment by kinverarity1** [comment]
Sounds great, but complex. Not sure if it is worth the code complexity for what seems like an unusual error?

**Comment by dcslagel** [comment]
Here is a simplified variation on the above approach.  Would this be a reasonable approach?  Is it better than the more complex approach?

1. Register when a section has been read.

2. If we are in the ~Curve section and see another section header before any data, check if that section header has been registered. 

3. If the new section has NOT been registered, then we treat the ~Curve section as empty and read following data as belonging to the new section.

4. If the new section has been registered then ignore this 2nd section header and read the following data as ~Curve section data

**Comment by JustinGOSSES** [comment]
Do I understand it right then that there is an implicit assumption that if there are two of the same sections then the headers are in first and data is in second?

**Comment by dcslagel** [comment]
It is different from that.  It is rather: if there are 2 of the same meta-data section headers (Version, Well, Parameter, Other, Curves), how should we handle the data that follows the 2nd/duplicate metadata section header?

**Comment by kinverarity1** [comment]
Approach 2 seems reasonable?

**Comment by dcslagel** [comment]
Okay,  I'll implement Approach-2 and post a pull-request of it for review.  It may take a few days...

**Comment by dcslagel** [comment]
Parsing of `260341080252801.19980915.NN.las` as outlined in `Approach-2:` above is implemented in: Handle duplicate required unique headers #71.

**Comment by dcslagel** [comment]
One of the problems with 261058081145201.19980227.ZE.las is that it changes the number of data columns from 5 to 6 in the ~ASCII section.  Example:
```
6.8	0	-1	0	0
6.9	0	-1	0	0
7	0	-1	122	0	-30.0
7.1	0	-1	122	0	-30.0
7.2	0	-1	122	0	-30.0
```

There are only 5 Curve names.

**Comment by dcslagel** [comment]
Pull Request #71 is updated to include a solution for the changing number of data columns found in `261058081145201.19980227.ZE.las`

Also revised the title to include this information:
`Handle duplicate headers and changing number of data columns` #71

**Comment by dcslagel** [comment]
Note: Removed the 'no-issue-activity' label because the bot uses that label to later trigger closing the issue and this issue is active and has a pending pull-request : #71

**Comment by JustinGOSSES** [comment]
Merged in #71 and put out new release. NPM now at v0.1.12

Thanks everyone.
